### PR TITLE
Delegate hyper-geometries to old ones for pre Kokkos-3.7

### PR DESCRIPTION
**Pull request by aprokop**
Per https://github.com/arborx/ArborX/pull/717#discussion_r961611361

### CODE REVIEW

File: src/geometry/ArborX_HyperBox.hpp [hunk 1]
```diff
@@ -98,6 +98,24 @@ struct Box
...
+
+#include <ArborX_Box.hpp>
+
+namespace ArborX::ExperimentalHyperGeometry
+{
+template <int DIM, class FloatingPoint = float>
+struct Box : public ArborX::Box
+{
+  static_assert(DIM == 3);
+  static_assert(std::is_same<FloatingPoint, float>{});
```
Comment (dalg24) [inline src/geometry/ArborX_HyperBox.hpp:111]:
```suggestion
  static_assert(std::is_same_v<FloatingPoint, float>);
```

File: src/geometry/ArborX_HyperPoint.hpp [hunk 2]
```diff
@@ -48,6 +48,27 @@ Point(T x, T y, T z) -> Point<3, T>;
...
+
+#include <ArborX_Point.hpp>
+
+namespace ArborX::ExperimentalHyperGeometry
+{
+template <int DIM, class FloatingPoint = float>
+struct Point : public ArborX::Point
+{
+  static_assert(DIM == 3);
+  static_assert(std::is_same<FloatingPoint, float>{});
```
Comment (dalg24) [inline src/geometry/ArborX_HyperPoint.hpp:61]:
```suggestion
  static_assert(std::is_same_v<FloatingPoint, float>);
```

File: src/geometry/ArborX_HyperSphere.hpp [hunk 3]
```diff
@@ -49,18 +49,34 @@ struct Sphere
...
+
+#include <ArborX_Sphere.hpp>
+
+namespace ArborX::ExperimentalHyperGeometry
+{
+template <int DIM, class FloatingPoint = float>
+struct Sphere : public ArborX::Sphere
+{
+  static_assert(DIM == 3);
+  static_assert(std::is_same<FloatingPoint, float>{});
```
Comment (dalg24) [inline src/geometry/ArborX_HyperSphere.hpp:62]:
```suggestion
  static_assert(std::is_same_v<FloatingPoint, float>);
```

**Comment by aprokop** [comment]
@dalg24 Thanks, I have to retrain my brain to start doing `std::is_same_v`.

**Comment by dalg24** [comment]
>Do you think we should introduce `GeometryTraits::dimension_v<Geometry>`?

Eventually something like it.  I still haven't gotten over my disliking both the nested namespace and the trait name.

**Comment by aprokop** [comment]
> Eventually something like it. I still haven't gotten over my disliking both the nested namespace and the trait name.

It does make it easier to write code. I think it's orthogonal to your disliking of the namespace and the trait name. If we decide to change those, we would also change `dimension_v`.

File: src/ArborX_LinearBVH.hpp [hunk 4]
```diff
@@ -191,7 +189,8 @@ BasicBoundingVolumeHierarchy<MemorySpace, BoundingVolume, Enable>::
   // map primitives from multidimensional domain to one-dimensional interval
   using LinearOrderingValueType = Kokkos::detected_t<
       Details::SpaceFillingCurveProjectionArchetypeExpression,
-      SpaceFillingCurve, Box, Point>;
+      SpaceFillingCurve, decltype(bbox),
+      std::decay_t<decltype(Access::get(primitives, 0))>>;
```
Comment (dalg24) [inline src/ArborX_LinearBVH.hpp:193]: Make sure you do `typename Primitives::value_type` in #729

**Review by dalg24 — APPROVED** [review]

**Comment by aprokop** [comment]
Merging without the HIP build, as it is going to timeout anyway due to having a single HIP runner for all Kokkos and ArborX builds.

**Pull request merged**